### ISSUE TITLE
Error when coping data to Cloudflare R2

### ISSUE THREAD

**Issue by djouallah** [duckdb/duckdb#5685]
### What happens?

using the S3 API, when I try to copy a file to R2, I get this error
`duckdb.IOException: IO Error: Unexpected response during S3 multipart upload finalization`

the same file was fine using AWS Boto package

### To Reproduce

```
import streamlit as st
import duckdb
con=duckdb.connect()

con.execute(f'''
install httpfs;
LOAD httpfs;
PRAGMA enable_object_cache ;
set s3_region = 'auto';
set s3_access_key_id = "{st.secrets["aws_access_key_id_secret"]}" ;
set s3_secret_access_key = '{st.secrets["aws_secret_access_key_secret"] }';
set s3_endpoint = '{st.secrets["endpoint_url_secret"]}'  ;
SET s3_url_style='path';
COPY (select * from 'Parquet/0.1/lineitem.parquet' ) to 's3://delta/1/lineitem.parquet'
''')
```

### OS:

windows 10

### DuckDB Version:

0.6.2.dev480

### DuckDB Client:

python

### Full Name:

mimoune djouallah

### Affiliation:

Personal project

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by ruchirj** [comment]
@samansmink I am running into the exact same issue. Were you able to identify the root cause?

**Comment by samansmink** [comment]
@ruchirj thanks for reporting this, I haven't had time yet to look into this due to the holidays, but its interesting to hear that you run into this in Minio, as we test all S3 code with Minio in our CI. I see however that we have a somewhat old version pinned. I'll try to find some time either this or next week to look into this

**Comment by ruchirj** [comment]
@samansmink I use the latest `0.6.1` version of the Java DuckDB library and the latest version of Minio. Which versions are you using because if this error goes away by back pinning it would help unblock me while you look into this and by the way appreciate that you are!

**Comment by samansmink** [comment]
@ruchirj our minio config for CI is here: https://github.com/duckdb/duckdb/blob/master/scripts/minio_s3.yml, so we're using minio/minio:RELEASE.2021-11-03T03-36-36Z currently

**Comment by pquadri** [comment]
same issue here, duckdb 0.7.0 and actual s3, was working with duckdb 0.6.1.

**Comment by complex64** [comment]
We've recently hit this issue and resolved it by setting the correct S3 permissions (i.e. allowed multipart operations).

**Comment by pquadri** [comment]
@complex64 we should have full permissions on the bucket already in place. Any quick way to check if we are missing some specific capability? (I checked through AWS console and aws cli, everything seemed fine)

**Comment by chrisbrain** [comment]
We are also seeing this now after trying to upgrade from 0.6.1 to 0.7.0.

The IAM user has AmazonS3FullAccess.

**Comment by Mause** [comment]
I'm working on adding improved http error reporting support, but in the meantime, if one of you could use Wireshark to look at the error you're experiencing, that would help us look into it

**Comment by tobilg** [comment]
I have the same issue with latest master and the correct permissions are set on the execution role according to https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html#mpuAndPermissions

**Comment by tobilg** [comment]
I compared v0.6.1 and v0.7.0, there didn't change much in the `FinalizeMultipartUpload` method:  
https://github.com/duckdb/duckdb/compare/v0.6.1...v0.7.0#diff-9ff786b314ba325d78b3dda594cb8d35bc53a57e8096febcc2c1c9bb13cd28efR405-R436

I don't think that this is the reason for this bug (not a C++ expert at all though)...

**Comment by Mause** [comment]
I think we've identified the issue here, relating to assumptions around periods not appearing in urls (aside from as the file extension). This assumption fails when periods appear in urls for other reasons (eg, presigned urls)

**Comment by Arttii** [comment]
Is there a workaround for this?

**Comment by Arttii** [comment]
@samansmink thanks!

**Comment by djouallah** [comment]
@samansmink is there a pr already in master ?

**Comment by samansmink** [comment]
working on it, pr is almost ready!

**Comment by Arttii** [comment]
I am also getting the same error when running something like this:

```sql
CREATE TABLE IF NOT EXISTS crawl_index (message VARCHAR, date DATE)";
insert into table values('yolo',current_date);
COPY table TO 's3://bucket/data' (FORMAT PARQUET, PARTITION_BY (date), ALLOW_OVERWRITE TRUE);
```

Do you think it's covered by your PR? Or this might be another issue? The error is exactly the same.

**Comment by tobilg** [comment]
As far as I understood the PR has the goal to fix this @Arttii

**Comment by tobilg** [comment]
It works now on latest master with S3, also the column order is now correct. Thanks a lot @samansmink / @Mytherin!

**Comment by djouallah** [comment]
@samansmink  still it does not works with Cloudflare.

**Comment by legout** [comment]
> @samansmink still it does not works with Cloudflare.

Did you try it with fsspec? That worked for me.

**Comment by djouallah** [comment]
Fsspec works fine with CloudFlare but there is no http metadata cache

**Comment by samansmink** [comment]
@djouallah I'll look into it!

**Comment by Mause** [comment]
@samansmink I wonder if the workaround for fsspec is possibly related to this issue? Apparently there's an additional ACL=private parameter that needs to be passed even for fsspec's s3fs

**Comment by elithrar** [comment]
We're effectively throwing here: https://github.com/duckdb/duckdb/blob/d8a1081030ca0cf7211483af01922b1d14eeebeb/extension/httpfs/s3fs.cpp#L439-L442

What I can't see/debug is what the raw response the CLI/s3fs is getting back from R2, but I suspect one of the following is the root cause:

* If the parts aren't equal (minus the last), then the multi-part upload will fail: https://github.com/duckdb/duckdb/blob/d8a1081030ca0cf7211483af01922b1d14eeebeb/extension/httpfs/s3fs.cpp#L419-L426
* Whether the AWS-centric response buffer is leading to the error being thrown: https://github.com/duckdb/duckdb/blob/d8a1081030ca0cf7211483af01922b1d14eeebeb/extension/httpfs/s3fs.cpp#L430-L431

A `--verbose` or `--log-http` flag that logs the requests & response bodies would be helpful here without having to jump through mitm-proxy/Charles/WireShark shenanigans.

**Comment by samansmink** [comment]
@elithrar thanks for the input, it was an url encoding issue in duckdb after all, we were incorrectly not encoding the multipart upload id in the query params

**Comment by djouallah** [comment]
@samansmink thanks, it works
